README.md
# Yoda

### Платформа для написания интеграционных тестов
> Небольшая платформа, которая призвана решать серьезные задачи.

<br />


**ВНИМАНИЕ**: Знакомьтесь с основными [изменениями](./CHANGELOG.md)!<br />
**ВНИМАНИЕ**: Помимо [стандартных](http://webdriver.io/api.html) команд имеются [кастомные](https://stash.mail.ru/projects/QA/repos/wdio-api/browse)
<br />

### Установка

Для работы с этим пакетом используйте [grunt-init-yoda](https://stash.mail.ru/projects/QA/repos/grunt-init-yoda/browse).


**git**

```
git clone ssh://git@stash.mail.ru:2222/qa/yoda.git
```

**nvm**

Для работы с этим пакетом требуется node версии не ниже 7.8.0! <br />
Поскольку обновить node и npm на всех машинах до требуемой версии весьма проблематично — используйте [nvm](https://github.com/creationix/nvm):

```bash
curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.31.1/install.sh | bash
source ~/.bash{rc,_profile}
nvm install 7
nvm use 7
nvm alias default 7
```

**npm**

```
npm install
```

**java**

Для запуска сервера потребуется [установить](http://www.oracle.com/technetwork/java/javase/downloads/index.html) Java

**jdk**

В случае возникновения следующей ошибки:

```
Exception in thread "main" java.lang.UnsupportedClassVersionError: org/openqa/grid/selenium/GridLauncher : Unsupported major.minor version 51.0
```

Обновите [JDK](http://www.oracle.com/technetwork/java/javase/downloads/index.html)

### Как держать версии в актуальном состоянии:
1. ставим глобально npm-check-updates
	```bash
	npm i -g npm-check-updates
	```

2. В конфиге `.ncurc.json` указаны версии которые жестко зафиксированы и  мы никогда не захотим их поднимать.

3. Интерактивно обновляем версии
	```bash
	ncu -i -u
	```
4. Делаем `npm install`


### Совместимость

Пакет совместим о всеми известными окружениями, однако работа с npm-хуками в Сygwin имеет некоторые ограничения на запуск файлов с правами на исполнение. Также в нем отсутствует команда source. Поэтому будьте бдительны :)


### Документация

Основная документация по этому проекту находится [здесь](http://yoda-doc.mail.cloud.devmail.ru) + есть старая [дока](https://confluence.mail.ru/pages/viewpage.action?pageId=97748321) Саши Абашкина. Однако для получение актуальной информации о всех изменения следите за файлом [CHANGELOG.md](./CHANGELOG.md).


### Конфигурирование

Основной файл, который отвечает за конфигурирование сервера и запуск тестов распологается в директории:

```
tests/config.js
```

Править этот файл не рекомендуется. Он нужен для запуска тестов удаленно. Для локального запуска тестов используйте конфиг:

```
tests/config.local.js
```

Дополнительную информацию о формате конфига и его опциях смотрите [здесь](http://webdriver.io/guide/getstarted/configuration.html).


### CI

Пример запуска тестов в среде непрерывной интеграции смотрите [здесь](http://win110.dev.mail.ru:8080/view/Yoda/job/yoda.suites).

О конфигурировании Jenkins смотри в [@qa/jenkins-utils](ssh://git@stash.mail.ru:2222/qa/jenkins-utils.git)

### Использование

Запуск selenium-standalone:

```
npm start
```

По умолчанию сервер запускается автоматически, но если вы все-таки хотите использовать ручной запуск, то для этого потребуется закомментировать поле `services: ['selenium-standalone'],` в вашем локальном конфиге (`config.local.js`).


#### Зупустить тесты конкретного тестового набора:

```
npm test -- --suite=login
```

`--suite` может принимать множество значений:

```
npm test -- --suite='login,compose'
```

#### Запуск тестов по фильтру:

```
npm test -- --suite=login --grep=TESTMAIL-8674
```

*Опция `--grep` принимает регулярное выражение*

```
npm test -- --suite=login --grep=TESTMAIL-867[45]
```

В этом случае, будут запущены два теста TESTMAIL-8674 и TESTMAIL-8675


#### Выполнить тесты на заданном адресе:

```
npm test -- --grep=TESTMAIL-8674 --baseUrl=https://e.mail.ru/login
```

#### Отладка тестов

Отлаживать тесты можно либо из консоли, либо из дебаггеров IDE, таких как WebStorm, VS Code, etc. Для запуска тестов с возможностью отладки, необходимо использовать опцию `--debug`:

```
npm test -- --grep=TESTMAIL-8674 --debug
```

По умолчанию используется порт 6666 (можно передать любой).

Рекомендуется отлаживать тесты по одиночке. При запуске тестов с включенной опцией `debug`, в консоли появится сообщение:

```
Debugger listening on 127.0.0.1:6666
```

Выполнение теста остановится на инструкции

```js
debugger;
```

Начать отладку пожно подключившись к процессу по заданному адресу. Для IDE WebStorm нужно выполнить:

```
Run -> Debug... -> Edit Configurations... -> Add New Configuration -> Node.js Remote Debug
```

Там необходимо выставить свои address и port и подключиться к отладчику.

Для VS Code необходимо следовать инструкциям из [статьи (medium.com)](https://medium.com/@dakshika/remote-debug-node-js-application-using-visual-studio-code-dc0fa0b4dec4#.3zk3yqg7h).


### Логи и отчеты

**cache/tests/logs** — логи сервера
**cache/tests/shots** — скриншоты с упавшими тестами
**cache/tests/reports** — отчеты о прогоне

### Линтер

Для валидации кода тестов мы используем tslint </br>

*Проверка запускается только на дифф*

```json
{
	"scripts": {
		"test": "yoda --lint"
	}
}
```

*Проверить все файлы (нерекомендуется — неоправданно долго)*

```json
{
	"scripts": {
		"test": "yoda --lint=."
	}
}
```

*Проверить заданный список файлов и директорий*

```json
{
	"scripts": {
		"test": "yoda --lint=pages --lint=cases --lint=steps"
	}
}
```

### API

Во всех тестах через степы доступны следующие методы:


#### Page#open({...})

Открытие требуемого представления

```ts
import Messages from '../../steps/messages';

describe(() => {
	it ('Открытие страницы списка писем', => {
		Messages.open();
	});
});
```

Если блок `describe` анонимный, то в качестве описания будет использоваться только название файла (которое соответствует номеру таска в JIRA).

ЗАПОМНИТЕ: писать номер таска в описании не нужно, он подставляется автоматически.

Метод auth дополнительно принимает параметры запроса:

```js
Messages.open({
	foo: 1
});
```

#### Page#features([...])

Включение фич:

```js
import Messages from '../../steps/messages';

describe(() => {
	beforeEach(() => {
		Messages.features([
			'check-missing-attach',
			'disable-ballons',
			'no-collectors-in-compose'
		]);

		Messages.open();
	});

	it ('Открытие страницы списка писем', => {
		// ...
	});
});
```

Используейте символ `:` если фиче требуется передать какое-то значение:

```js
Messages.features([
	'check-missing-attach:1'
])
```

#### Page#register(type=basic, { username, password, ... })

Регистрация нового пользователя

```js
import Messages from '../../steps/messages';

describe(() => {
	let { username, password } = Messages.register('basic');

	Messages.auth('basic', { username, password });
	Messages.open();
});
```

#### Page#auth(type=basic, { username, password, ... })

Авторизация

```js
import Messages from '../../steps/messages';

describe(() => {
	Messages.auth();
	Messages.open();
});
```

Метод auth дополнительно принимает тип авторизации:

```js
Messages.auth('external');
```

Список доступных типов: pdd, external, basic (используется по умолчанию)

Также есть возможность авторизоваться конкретным пользователем:

```js
Messages.auth('basic', {
	username: 'aziza.voronova.52@list.ru',
	password: 'dknMuWpuzvG9'
});
```

Метод .credentials примает те же типы, что Page\#auth

#### store/authorization.credentials(type=basic, { id, login, domain, type }, timeout)

Получение авторизационных данные указанного типа

```js
import authorization from '@qa/yoda/store/authorization';

authorization.credentials('external', {
	domain: 'gmail.com'
});
```

Метод .credentials принимает те же типы, что Page\#auth

ВНИМАНИЕ: обязательно вызываейте этот метод `discard` для осовобождения занимаемого аккаунта!

#### store/authorization/accounts.get

Возвращает учетную запись из локального хранилища

ВНИМАНИЕ: Данные учетные записи разрешается использтвать только в тестах, которые не изменяют состояние аккаунта (например, авторизация). Во всех остальных случаях — используейте store/authorization.credentials

```js
import accounts '@qa/yoda/store/authorization/accounts';

accounts.get('gmail.com', ['pdd']);
```

Второй параметр опциональный и, как правило, имеет смысл только с некоторыми значениями, например `pdd`.


#### store/authorization/providers

Здесь содержится набор методов для получения списка провайдеров с учетом различных фильтров, в т.ч. и топа.


#### store/authorization/passwords

Позволяет получить предопределенный пароль заданой сложности


#### utils/account.passwordStrength

Позволяет определить сложность пароля


### Методы регрессионого тестирования средставами визуального сравнения

Убедитесь, что в вашем package.json-файле имеются следующие зависимости:

```json
{
	"dependencies": {
		"wdio-screenshot": "^0.2.3",
		"wdio-visual-regression-service": "^0.4.0"
	}
}
```

#### API

```
 .compareDocument([options])
 .compareViewport([options])
 .compareElement(locator, [options])
```

Доступные опции:

```
options.hide {string[]}              Скрывает заданные элементы
options.remove {string[]}            Удаляет заданные элементы
options.widths {number[]}            Задает размер изображениям (desktop)
options.orientations {number[]}      Устанавливает ориентацию (mobile)
options.misMatchTolerance {number}   Задает границы поиска несоотвествий (от 0 до 100)
options.viewportChangePause {number} Устанавливает время ожидания после
                                     изменения раземеров вьюпорта
```

Пример использования в проекте themes.mail.ru:

**config.js**

```ts
import * as VisualRegressionCompare from 'wdio-visual-regression-service/compare';

{
	/** Конфигурация для сервиса visual-regression */
	visualRegression: {
		compare: new VisualRegressionCompare.LocalCompare({
			referenceName : support.screenshot('./store/shots/expected'),
			screenshotName: support.screenshot('./store/shots/actual'),
			diffName      : support.screenshot('./store/shots/diff')
		})
	},

	/**
	 * Список сервисов, которые будет использоваться для запуска тестов
	 *
	 * selenium-standalone — позволяет автоматически запускать и останавливать
	 * работу selenium-сервера
	 */
	services: [
		/**
		 * Сервис для регрессионого тестирования через сравнение скриншотов
		 */
		'visual-regression'
	]
}
```

**cases/messages/TESTMAIL-34047.js**

```js
import Messages from '../../steps/messages';
import { dimensions, config } from '../../store';

let TIMEOUT = 30 * (60 * 1000);
let messages = new Messages();

describe('Темы. Общее соответствие оформления на списке писем', function () {
	this.timeout(TIMEOUT);

	before(() => {
		Messages.auth();
		Messages.open();
	});

	for (let theme in config) {
		for (let dimension of dimensions) {
			let { width, height } = dimension;

			it(`${theme}-${width}x${height}`, () => {
				messages.setTheme(theme);
				messages.setViewportSize(dimension);
				messages.compareDocument();
			});
		}
	}
});
```

**cases/messages/index.js**

```ts
import * as assert from 'assert'

import MailMessagesSteps from '@qa/yoda-e.mail.ru/steps/messages';
import MessagesPage from '../../pages/messages';

let page = new MessagesPage();

/** Модуль для работы с шагами представления */
class MessagesSteps extends MailMessagesSteps {
	constructor () {
		super();
	}

	/**
	 * Возвращает ссылку на инстанс страницы
	 *
	 * @type {Object}
	 */
	static get page () {
		return page;
	}

	compareDocument () {
		let { hide } = page.locators;

		super.compareDocument({ hide });
	}
}

export default MessagesSteps;
```

**pages/messages/index.js**

```js
import MailMessagesPages from '@qa/yoda-e.mail.ru/pages/messages';

/** Модуль для работы с представлением */
class MessagesPage extends MailMessagesPages {
	/**
	 * Локаторы
	 *
	 * @type {Object}
	 */
	get locators () {
		return this.extend(super.locators, {
			/** Список интерактивных элементов, которые требуется исключить */
			hide: [
				'#leftcol-banners',
				'.b-datalist__head',
				'.pm-menu__center__inner',
				'#portal-headline',
				'#SeptimaFeedback',
				'.footer__theme-widget',
				'#footer__portal'
			]
		});
	}
}

export default MessagesPage;
```

### Асинхронные тесты

Пример асинхронного теста:

```js
browser.timeouts('script', TIMEOUT);

let { value } = browser.executeAsync(function (name, value, resolve) {
	require(['features'], function (features) {
		var actual = features.use(name, value);

		resolve(actual);
	});
}, name, value);
```

А что если нужно вернуть результат объекта Promise?

```js
let value  = browser.waitForPromise(() => {
	return new Promise((resolve, reject) => {
		resolve(true);
	});
}, TIMEOUT, 'Error');
```

Обратите внимание, что мы использовали кастомную команду `browser.waitForPromise`, которая на вход получает колбек возвращающий промис, либо сам промис.


Этот же пример можно переписать следующим образом:

```js
let { value } = browser.call(() => {
	return browser.executeAsync(function (name, value, resolve) {
		require(['features'], function (features) {
			var actual = features.use(name, value);

			resolve(actual);
		});
	}, name, value);
});
```

Однако в этом случае, мы не сможем задать сообщение об ошибке и таймаут!
`browser.timeouts` и `browser.timeoutsAsyncScript` также не помогут в этом вопросе.

Есть также распространенная практика использовать асинхронный `waitUntil`:

```js
browser.waitUntil(function async () {
	return Promise.resolve(true);
});
```

Проблемы такого способа:

* `waitUntil` не вернет результат, поспольку всегда ожидает получить `true`
* Если забыть указать название функции `async`, то метод будет работать синхронно
* Опция timeout будет проигнорирована до тех пока не будет вызван хотя бы один раз `reject`!
* Без [фиктивного](https://github.com/webdriverio/webdriverio/issues/1407) `reject`'a тест зависнет навечно.

Однако, если вам нужно дождаться какого-то состояния интерфейса, то waitUntil отличный помощник:

```js
browser.waitUntil(function async () {
	return browse.isVisible('body');
});
```

### Файловая система

Для работы с файлами, которые хранятся на диске используйте системное хранилище:

```js
import system from '@qa/yoda/store/system';

system.file('имя_файла');
```

### Структура проекта

Ниже будет рассмотрена структура проекта на примере e.mail.ru:

```
➜ tests
	➜ cases
		➜ <название_раздела>
			<название_кейса>.js
	➜ pages
		➜ <название_раздела>
			<название_предствавления>.js
	➜ steps
		➜ <название_раздела>
			<шаги_раздела>.js
	➜ store
		➜ <название_раздела>
			<хранилище_раздела>.js
	➜ utils
		➜ <название_раздела>
			<утилиты_раздела>.js
	config.js
	config.local.js
```


Пример:

```
cases
	compose
		MAIL-15143
			TESTMAIL-18932.js

pages
	compose
		attaches.js

steps
	compose
		attaches.js

store
	compose
		attaches.js

tasks
	yoda.js

utils
	attaches
		add.js
```

| Сущность  | Назначение              | Взаимодействие |
|-----------|-------------------------|----------------|
| **cases** | Тест-кейсы              | store, steps
| **pages** | Элементы предстравления | store, browser
| **steps** | Шаги                    | store, pages, steps
| **store** | Хранилище               | store
| **tasks** | Задания                 | tasks
| **utils** | Утилиты                 | store

Обратие внимание, что взаимодействие между сущностями строго регламентировано. Например, это значит, что вы не можете внутри какого-либо тест-кейса обращься к элементам представления (все объекты представления включая `browser` можно вызывать только в директории `pages` и нигде больше).


#### cases

Пример тест-кейса:

```js
import * as assert from 'assert'

import login from '../../steps/login';
import form from '../../steps/login/form';

describe(() => {
	it('Проверка отображения элементов на форме авторизации', () => {
		login.open();

		form.checkDefaultDomain();
		form.checkTitle();
	});
});
```

Пример теста с авторизацией:

```js
import Messages from '../../steps/messages';

describe(() => {
	it('Проверка перехода на страницу списка писем.', () => {
		Messages.auth();
		Messages.open();
	});
});
```

В тест-кейсе вызываются **только** шаги!


#### pages

Пример представления:

```js
import * as assert from 'assert'
import PageObject from '../../pages';

class Login extends PageObject {
	constructor () {
		super();
	}

	get location () {
		return '/login'
	}

	get locators () {
		return {
			container: '.login-page__external'
		}
	}
}

export default Login;
```

#### steps

Пример степов:

```js
import * as assert from 'assert'
import providers from '@qa/yoda/store/authorization/providers';

import Steps from '../../steps';
import LoginForm from '../../pages/login/form';

class LoginFormSteps extends Steps {
	constructor () {
		super();

		this.loginForm = new LoginForm();
	}

	/**
	 * Проверить домен, который используется по умолчанию
	 *
	 * @param {string} provider
	 */
	checkDefaultDomain (provider) {
		this.loginForm.getActiveDomain('mail.ru');
	}

	/**
	 * Сверить активный домен
	 *
	 * @param {string} provider
	 */
	getActiveDomain (provider) {
		assert.equal(this.loginForm.activeDomain, provider,
			`Передан неверный провайдер ${provider}`);
	}

	/**
	 * Получить заголовок формы
	 *
	 * @returns {string}
	 */
	checkTitle () {
		assert.equal(this.loginForm.title, 'Вход в почту',
			'Не удалось проверить заголовок формы');
	}
}

export default LoginFormSteps;
```

#### store

Пример хранилища данных:

```
➜ store
	➜ authorization
		index.js
		providers.js

	➜ login
		providers.js
```

**store/authorization/providers.js**

```js
import Store from '../../store';

/** Модуль для работы с данными почтовых провайдеров */
export default {
	/**
	 * Cписок провайдеров
	 *
	 * @returns {Array}
	 */
	list: [
		{
			name: 'mail.ru',
			type: 'internal',
			data: [
				'mail.ru',
				'mail.ua',
				'inbox.ru',
				'list.ru',
				'bk.ru'
			]
		}
	],

	/**
	 * Добаавляет список провайдеров
	 *
	 * @param {...Array} providers — список провайдеров
	 */
	set (...providers) {
		this.list.push(...providers)
	}

	/**
	 * Возвращает список провайдеров
	 *
	 * @param {Array|null} providers — заданный список провайдеров
	 * @returns {Array}
	 */
	get (providers) {
		if (providers) {
			return this.list.filter(provider => {
				if (providers.includes(provider.name)) {
					return provider;
				}
			});
		}

		return this.list;
	}
}
```

**store/login/providers.js**

```js
import authProviders from '@qa/yoda/store/authorization/providers';

/** Модуль для работы с данными почтовых провайдеров на странице логина */
export default {
	/**
	 * Получить активный список провайдеров (пиктограммы)
	 *
	 * @type {Array}
	 */
	get active () {
		return authProviders.get([
			'mail.ru',
			'yandex.ru',
			'rambler.ru',
			'gmail.com'
		]);
	},

	/**
	 * Получить список провайдеров (селект)
	 *
	 * @type {Array}
	 */
	get select () {
		return authProviders.get([
			'mail.ru',
			'yandex.ru',
			'rambler.ru',
			'gmail.com',
			'yahoo.com',
			'hotmail.com',
			'outlook.com'
		]);
	}
};
```

### Требования

* Не обращайтесь в pages к объекту browser напрямую. Вместо этого используйте ссылку `this.page`.
* Устаревшие тесты, которые подлежат удалению зафиксируйте статусом Archive в JIRA.
* Писать номер таска в блоке `describe` не нужно, он подставляется автоматически.
* Универсальные команды добавляйте в проект [@qa/wdio-api](ssh://git@stash.mail.ru:2222/qa/wdio-api.git)
* Используйте JSDoc аннотацию для документирования функций, которые принимают параметры и возвращают значения.
* Все файлы, в которых определены классы должны возвращать ссылки, а не инстанс.
* Всегда определяйте `location` и `locators.container` в индексном файле вашего предствления (page).
* Всегда определяйте `page` в индексном файле степов.
* Не используйте сокращения вида err, dfd, fn, и пр.
* Для переменной, которая сохраняет состояние используйте название `actual`.
* Прижерживайтесь существующей структуры и организации кода проекта.
* Для работы с любыми данными используйте хранилища (`store`).
* Если вы работаете с полями формы, то у вас должны быть определены как минимум следующие типы методов:


| Метод  | Назначение                                | Определение    |
|---------------------|------------------------------|----------------|
| **getField**        | Получить значение элемента   | pages
| **getFieldValue**   | Получить значение элемента   | pages
| **setFieldValue**   | Установить значение элемента | pages, steps
| **clearFieldValue** | Очистить значение элемента   | pages, steps


Необязательные методы:

| Метод              | Назначение                   | Определение    |
|--------------------|------------------------------|----------------|
| **showField**      | Показать элемент             | pages, steps
| **hideField**      | Скрыть элемент               | pages, steps
| **isVisibleField** | Проверить видимость элемента | pages, steps
| **clickField**     | Сделать клик в поле          | pages, steps
| **tabField**       | Перейти в поле по табу       | pages, steps


Эти методы должны примать имена полей и значения:

**<page>**

```js
{
	/**
	 * Получить элемент поля по имени
	 *
	 * @param {string} name — имя поля.
	 * Доступные значения (from, to, cc, bcc, subject, priority, receipt, remind)
	 *
	 * @returns {WebDriver}
	 */
	getField (name) {
		return this.page.element(this.locators.fields[name]);
	}
}
```

```js
{
	/**
	 * Получить значение поля по имени
	 *
	 * @see getField
	 * @param {string} name — имя поля
	 * @returns {string}
	 */
	getFieldValue (name, value) {
		return this.getField(name).getValue();
	}
}
```

**<step>**

```js
{
	/**
	 * Получить значение поля по имени
	 *
	 * @see form.getField
	 * @param {string} name — имя поля
	 */
	getFieldValue (name) {
		form.getFieldValue(name);
	}
}
```

**Универсальные локаторы**

Да, локаторы можно сделать универсальными:

```js
{
	get locators () {
		let container = 'body';

		return {
			button: new Proxy({}, {
				get (target, name) {
					return `${container} .b-btn-social__social__link_${name}`;
				}
			})
		}
	}
}
```

Далее к такому локатору можно обращаться так:

```
this.locators.button.vk
```

Так вы получите такой локатор:

```
body .b-btn-social__social__link_vk
```


CHANGELOG.md
# Changelog
## 9.16
* Добавили апишные методы для клинмастера, для имитации оплаты
* Добавили города для тестирования личных настроек и профиля на id.mail.ru
* Добавили фиксированные коды у телефонов
* Поправлены ссылки на спеку по webauthn-у

## 9.15
* Добавили тестовые методы SWA
* Починили метод с добавлением тестовой подписки на пуши

## 9.14.1
* Настроили поход за кукой ftrs в тестах

## 9.14.1
* Добавлены методы WaitForUrl, CheckQueryParams, HasNoQueryParams в классе Browser
* Добавлены методы CheckFocus и WaitForFocus в классе Element 

## 9.14.0
* Обновили версию chromedriver до 78
* Методы для очистки и редактирования папки, для изменения разрема ящика

## 9.13.5
* Исправлен CAPTCHA_CRACKER_URL в конфиге

## 9.13.4
* Добавлены новые методы api
* Исправлены баги в авторизации и работе с mail-api методами

## 9.12.0
* Версия ts 3.6.3, поправлены тайпинги

## 9.11.0
* Добавлены методы для добавления ключей WebAuthn
* Добавлен тстовый метод для включения на аккаунте входа/восстановления через пуш
* Перенесли @qa/rpc в йоду


## 9.10.3
* Добавлены методы для добавления сборщиков

## 9.10.2
* В зависимости добавлены chromedriver и webdriverio
* Изменен адрес сервиса для авторизации ящиками test_login

## 9.10.0
* Новые методы для работы с gibdd api
* Доработки в Mail API и Cloud API

## 9.9.4
* Исправлен баг в lib/generator
* Добавлены методы SwitchParentFrame и KeysPress в классе Browser
* @deprecated Метод Element.keyPress
* @qa/transport": "2.3.2
* Доработки степов toggleThreads и waitForLetterBySubjectInStatus
* Доработка utils/assert-defined

## 9.9.2
* Добавлена возможностьавторизовываться в тестах аккаунтами типа test_login
* Немного доработок по методам для шаринга ящика (Glasha)
* Первые общие элементы и методы - элементы порталки.
* Element теперь дженерик

## 9.9.1
* Починили баг в методе `open`

```
null is undefined
[chrome #0-0] TypeError: null is undefined
[chrome #0-0]     at Object.assertDefinedValue
...
```
* Добалены mail-api методы для шаринга ящика (Glasha)


## 9.9.0

* Строгие проверки типов
* Lock-файл
* Починили баг в генерилке элементов, из-за которого в d.ts появлялось: `import("/Users/anatolijostapenko/repos/yoda/element/index").Element`
* Новые методы у элементов: `Element.RightClickTo`
* Новые internal и mailapi методы

## 9.8.6

* Добавили сервис @qa/yoda/wdio/mocha-hooks, который служит заменой пакета @qa/wdio-mocha-hooks

## 9.5.0

* Новые методы для работы с авторизацией в [utils/authorization.ts](utils/authorization.ts)
* Новый базовый класс в концепции Yoda 2.0 [Page](element/page.ts)

## 9.4.26

* Обновил реализацию метода очищения инпута, чтобы он работал и с react-инпутами

## 8.0.25

* Обновили canary-куку

## 8.0.24

* Поле raw в транспорте было заменено на rawFilename

## 8.0.23

* Поднята версия @qa/transport

## 8.0.22

* Удаалена директория commands

## 8.0.21

* В опции добавлено поле  hostname
* Обновлен @aa/account-manager 

## 8.0.20

* Исправлено падание для --suite=axis:default

## 8.0.19

* Удалено выставление qa-canary

## 8.0.18

* Добавлено выставление qa-canary

## 8.0.17

* Добавлено выставление qa-куки для страниц, которые не используют авторизацию

## 8.0.16

* Обновлена версия @qa/wdio-utils

## 8.0.15

* Добавлена пропущенная зависимомть minimist

## 8.0.14

* Добавлен альтернативный экспорт констант в utils/constants

### 8.0.13

* Обновлен @aa/account-manager

## 8.0.12

* Обновлен список зависимостей

## 8.0.11

* Более информативное описание для методов, которые работают со скриншотами

## 8.0.10

* Изменение сообщение об ошибке

## 8.0.9

* В поле content интерфейса IPutMessageData убран тип Buffer

## 8.0.8

* Обновлен пакет @qa/transport

## 8.0.7

* Исправлен метод auth
* Переменная NAME в PageObject переименована в name

## 8.0.6

* Исправлен метод session

## 8.0.5

* Добавлена переменная NAME в PageObject

## 8.0.4

* Обновлен пакет @qa/wdio-utils

## 8.0.3

* Обновлен пакет @qa/transport

## 8.0.2

* Исправлен метод wait

## 8.0.1

* Исправлена ошибка с tig

## 8.0.0

* Модули tir и wdio-yoda-api-service обновлены, а его декларации типов удалены
* Удален файл с декларациями типов для process.debugPort
* Удалена заглушка tasks/globals для deprecated
* В модулях index.ts, tasks/runner.ts и store/authorization/index.ts Promise заменен на async/await
* Удалена директория hooks
* Исправлено описание

## 7.3.14-7.3.15

* Исправлен интерфейс по работе со скриншотами

## 7.3.13

* Поднята версия @qa/transport

## 7.3.12

* Поднята версия @qa/wdio-utils

## 7.3.11

* Исправлено описание методов refresh и wait

## 7.3.10

* Добавлено описание для waitForClass

## 7.3.9

* Поднята версия @qa/transport

## 7.3.8

* Исправлена проблема с открытием страницы без предварительной авторизации

## 7.3.7

* Модуль wdio-yoda-api-service перенесен в peerDependencies

## 7.3.6

* Поднята версия @qa/transport

## 7.3.5

* Поднята версия @qa/transport
* Улучшена совместимость с typescript@2.3.2

## 7.3.4

* Поднята версия @qa/transport

## 7.3.3

* Исправлено поле raw в putMessage

## 7.3.2

* Добавлены интерфейсы для версия getTagName, getTagNames, getValue, getValues

## 7.3.1

* Поднята версия @qa/transport

## 7.3.0

* Поднята версия @qa/transport
* Добавлена покладка треда через utils/transport/putThread

## 7.2.4

* Добавлено логирование для кук

## 7.2.3

* Актуализирован интерфейс транспорта
* Добавлены интерфейсы для getText и getTexts

## 7.2.2

* Поднятие версии wdio-yoda-allure-utils-service

## 7.2.1

* Добавлено описание для методов features и inject

## 7.2.0

* Добавлен транспорт писем `utils/transport`
* Добавлено расширенное описание для методов open и auth

## 7.1.16

* Добавлено описание для метода auth
* session теперь возвращает Credentials, а не boolean

## 7.1.15

* Убраны дубликаты степов
* Исправлена опечатка в описанании метода open

## 7.1.14

* Добавлены описания степов

## 7.1.12-13

* Поднята версия @qa/account-manager

## 7.1.11

* Удалены файлы .eslintignore и .eslintrc.js

## 7.1.9-10


config.js
'use strict';

let AccountManager = require('@qa/account-manager'),
	WebDriverAPI = require('@qa/wdio-api-mail.ru'),
	TestTools = require('@qa/test-tools'),
	capabilities = require('@qa/wd-capabilities');

let account = new AccountManager.Hooks();
let support = new TestTools.Support();


/** @namespace browser **/
exports.config = {
	/*
	 * Настройки с которыми запущен сервер Selenium
	 *
	 * Операционная система: Windows
	 * Список доступных браузеров: Chrome, Firefox, Opera, IE 11
	 * Грид: http://vagabond3.dev.mail.ru:4444/grid/console
	 *
	 * Операционная система: Linux
	 * Список доступных браузеров: Chrome, Firefox
	 * Грид: http://win110.dev.mail.ru:4444/grid/console
	 */
	host: 'localhost',
	port: 4444,
	path: '/wd/hub',

	/* Базовый адрес тестирования */
	baseUrl: 'https://e.mail.ru',

	/* Доступные значения: silent, verbose, command, data, result, error */
	logLevel: 'silent',

	/*
	 * Максимальное время на выполнение команды.
	 * Если какая-то из команд фреймворка не получит за это время результат,
	 * то выполнение тестов будет прервано.
	 */
	waitforTimeout: 30 * 1000,

	/* Максимальное время на выполнение повторного запроса. */
	connectionRetryTimeout: 10 * 1000,

	/* Количество инстансов параллельного запуска тестов */
	maxInstances: 1,

	/** Использовать синхронное API */
	// sync: true,

	/*
	 * Опция позволяет отладчику остановить выполнение тестов
	 * в месте вызова инструкции debugger.
	 * Для использования этой опции требуется наличие пакета node-inspector
	 */
	debug: false,

	/* Доступные значения: cucumber, mocha, jasmine */
	framework: 'mocha',

	mochaOpts: {
		ui: 'bdd',
		retries: 2
	},

	/* Для реппортера Allure требуется наличие установленного плагина в CI */
	reporters: ['dot', 'junit'],

	reporterOptions: {
		outputDir: './cache/tests/reports'
	},

	/* Директория, куда будут складываться скриншоты */
	// screenshotPath: './cache/tests/shots',

	/* Директория, куда будут складываться логи */
	logfile: './cache/tests/logs',

	/*
	 * Список файлов с тестами.
	 * Порядок файлов сохраняется, дубликаты исключаются
	 */

	/*
	 * Список файлов, которые требуется запустить
	 * Этот набор запускается всегда независимо от секции <suite>
	 */
	// specs: [ ],

	/**
	 * Набор тестовых кейсов
	 *
	 * { <suite>: [ <files> ] }
	 */
	suites: support.suites('tests'),

	/*
	 * Обратие внимание на то, что браузеры запускаются параллельно
	 * Конфигуратор https://wiki.saucelabs.com/display/DOCS/Platform+Configurator
	 *
	 * Внутри каждой группы доступны поля specs и exclude
	 *
	 * Список стандартных опций WebDriver и пр. см ниже по ссылке:
	 * @see https://stash.mail.ru/projects/QA/repos/wd-capabilities/browse
	 */
	capabilities: [
		capabilities.get('chrome')
	],

	before (capabilities, specs) {
		let commands = new WebDriverAPI();

		commands.export('all');
	},

	beforeSuite () {
		return account.session();
	},

	afterSuite () {
		return account.discard();
	}
};
